### ISSUE TITLE
SQL Server upsert

### ISSUE THREAD

**Issue by r2evans** [ankane/dbx#21]
I've done a bit of work getting similar code to work between different DBMSes, including Upsert in both postgres and sqlserver.

While in Postgres an upsert might look like this:

```r
INSERT INTO tblname ( nm1 , nm2 )
SELECT nm1 , nm2 FROM tmptable
ON CONFLICT ( on1 , on2 ) DO update set nm1=... 
```

It may look like this in sql server,

```r
MERGE tblname WITH (HOLDLOCK) as tgt
USING (SELECT nm1,nm2 FROM tmptable) as src
  ON ( tgt.on1=src.on1 and tgt.on2=src.on2 )
WHEN MATCHED THEN update set nm1=... 
WHEN NOT MATCHED THEN insert (nm1,nm2) values (...) ; 
```

If the "matched" action is instead to "do nothing" (as in: *insert if not present, do not update if present*), then

```r
DECLARE @dummy int;
MERGE tblname WITH (HOLDLOCK) as tgt
USING (SELECT nm1,nm2 FROM tmptable) as src
  ON ( tgt.on1=src.on1 and tgt.on2=src.on2 )
WHEN MATCHED THEN UPDATE SET @dummy=1
WHEN NOT MATCHED THEN insert (nm1,nm2) values (...) ; 
```

It seems hackish, but I haven't found a more elegant way without many more checks.

This has been fairly performant for me (not a big bottleneck), and I've been using it on tables with 100M rows (with GUIDs as primary key).

Are you able to use this to include SQL Server support in upsert?

**Comment by ankane** [comment]
Hey @r2evans, thanks for the great info! It's not something I have planned, but if you decide to fork and implement, please share.

**Comment by ankane** [comment]
Just fyi, this is now available in 0.3.1 (#35).
